src/azure-cli-core/azure/cli/core/commands/progress.py

```diff
@@ -0,0 +1,176 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+from __future__ import division
+import sys
+from enum import Enum
+
+import humanfriendly
+
+BAR_LEN = 70
+
+
+class ProgressType(Enum):  # pylint: disable=too-few-public-methods
+    """ the types of progress """
+    Determinate = 0
+    Indeterminate = 1
+    # for views that want to accept both determinate and indeterminate progres
+    Both = 2
+
+
+class ProgressViewBase(object):
+    """ a view base for progress reporting """
+    def __init__(self, out, progress_type, format_percent=None):
+        self.out = out
+        assert isinstance(progress_type, ProgressType)
+        self.progress_type = progress_type
+        self.format_percent = format_percent
+
+    def write(self, args):
+        """ writes the progress """
+        pass
+
+    def flush(self):
+        """ flushes the message out the pipeline"""
+        self.out.flush()
+
+    def get_type(self):
+        """ returns the progress type """
+        return self.progress_type
+
+
+class DetProgressReporter(object):
+    """ generic progress reporter """
+    def __init__(self, total_value=1):
+        self.message = ''
+        self.curr_val = 0 if total_value else None
+        self.total_val = total_value
+
+    def add(self, kwargs):
+        """
+        adds a progress report
+        :param kwargs: dictionary containing 'message', 'total_val', 'value'
+        """
+        message = kwargs.get('message', '')
+        total_val = kwargs.get('total_val', 1.0)
+        value = kwargs.get('value', 0.0)
+        assert value >= 0 and value <= total_val and total_val >= 0
+        self.total_val = total_val
+        self.curr_val = value
+        self.message = message
+
+    def report(self):
+        """ report the progress """
+        percent = self.curr_val / self.total_val
```

Comment:
What happens if self.total_val is zero? The assert in add checks that it is greater than or equals to zero, which I assume implies that zero is a valid value? :)